File: ImagePickerMaster/src/main/java/com/example/apple/glidetest/FolderPopup.java
package com.example.apple.glidetest;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.PopupWindow;
import android.widget.TextView;
import com.example.apple.glidetest.bean.Folder;
import com.example.apple.glidetest.bean.FolderProvider;
import com.example.apple.glidetest.utils.GlideUtils;
import com.example.apple.glidetest.utils.OsUtils;
import java.io.File;

/**
 * Created by Apple on 17/5/31.
 */

public class FolderPopup {

    ListView lvFolder;
    TextView tvFolderShow;
    private PopupWindow popupWindow;
    private FolderProvider folderProvider;

    public FolderPopup(Context context, final OnFolderSelectListener listener) {
        View view = View.inflate(context, R.layout.folder_popup, null);
        view.findViewById(R.id.tv_folder_show).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                tvFolderShow.setSelected(!tvFolderShow.isSelected());
                popupWindow.dismiss();
            }
        });
        lvFolder = (ListView) view.findViewById(R.id.lv_folder);
        tvFolderShow = (TextView) view.findViewById(R.id.tv_folder_show);

        int height = OsUtils.getScreenHeight(context) - OsUtils.dp2px(context, 100) - OsUtils.getStatusBarHeight(context);
        popupWindow = new PopupWindow(view, ViewGroup.LayoutParams.MATCH_PARENT, height);
        folderProvider = FolderProvider.getInstance();
        popupWindow.setBackgroundDrawable(new BitmapDrawable(context.getResources(), (Bitmap) null));
        popupWindow.setAnimationStyle(R.style.popup_anim);
        lvFolder.setAdapter(new MyAdapter());
        tvFolderShow.setText(folderProvider.getSelectedFolder().name);
        tvFolderShow.setSelected(true);
        lvFolder.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                folderProvider.setSelectedFolder(folderProvider.getFolders().get(position));
                tvFolderShow.setText(folderProvider.getSelectedFolder().name);
                if (listener != null) {
                    listener.onFolderSelected();
                }
                popupWindow.dismiss();
            }
        });
        popupWindow.setOnDismissListener(new PopupWindow.OnDismissListener() {
            @Override
            public void onDismiss() {
                if (listener != null) {
                    listener.onDismiss();
                }
            }
        });
    }

    class MyAdapter extends BaseAdapter {
        @Override
        public int getCount() {
            return folderProvider.getCount();
        }

        @Override
        public Folder getItem(int position) {
            return folderProvider.getFolders().get(position);
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            ViewHolder holder;
            if (convertView == null) {
                convertView = View.inflate(lvFolder.getContext(), R.layout.item_folder_popup, null);
                holder = new ViewHolder(convertView);
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }
            GlideUtils.loadImage(new File(getItem(position).firstImagePath), lvFolder.getContext(), holder.ivFolder);
            holder.tvFolder.setText(getItem(position).name + " (" + getItem(position).getCount() + ")");
            return convertView;
        }
    }

    static class ViewHolder {
        ImageView ivFolder ;
        TextView tvFolder;

        ViewHolder(View view) {
            ivFolder= (ImageView) view.findViewById(R.id.iv_folder);
            tvFolder= (TextView) view.findViewById(R.id.tv_folder);
        }
    }

    public void showPopup(View parent) {
        popupWindow.showAtLocation(parent, Gravity.TOP, 0, 0);
    }

    public interface OnFolderSelectListener {
        void onFolderSelected();

        void onDismiss();
    }
}

File: ImagePickerMaster/src/main/java/com/example/apple/glidetest/FolderPopup.kt
package com.example.apple.glidetest

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.PopupWindow
import com.example.apple.glidetest.bean.FolderProvider
import com.example.apple.glidetest.utils.getView
import com.example.apple.glidetest.utils.loadImage
import kotlinx.android.synthetic.main.item_folder.view.*
import kotlinx.android.synthetic.main.popup_folder_popup.view.*
import kotlinx.android.synthetic.main.title_bar.view.*
import java.io.File

/**
 * Created by Apple on 17/8/23.
 */

class FolderPopup(private val context: Context) {

    private val popupWindow: PopupWindow
    private val folders = FolderProvider.instance.folders
    private var listener: OnFolderSelectedListener? = null
    private var contentView: View? = null

    init {
        contentView = View.inflate(context, R.layout.popup_folder_popup, null)
//        var height = context.getScreenHeight() - context.getStatusBarHeight() - context.dp2px(50f)
        popupWindow = PopupWindow(contentView, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, true)
        popupWindow.setBackgroundDrawable(ColorDrawable(Color.parseColor("#ffffff")))
        contentView!!.recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        contentView!!.recyclerView.adapter = FolderAdapter()
        popupWindow.animationStyle = R.style.popup_folder_anim
        contentView!!.tvCenter.isSelected = true
        contentView!!.tvCenter.setOnClickListener {
            popupWindow.dismiss()
        }
    }


    fun show(parent: View, listener: OnFolderSelectedListener) {
        this.listener = listener
        popupWindow.showAtLocation(parent, Gravity.NO_GRAVITY, 0, 0)
    }

    inner class FolderAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
        override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int) {
            if (holder is FolderHolder) {
                if (folders.get(position).firstImagePath != null)
                    loadImage(File(folders.get(position).firstImagePath), holder.itemView.ivFolder)
                holder.itemView.tvFolder.text = folders.get(position).name
                holder.itemView.tvCount.text = "(" + folders.get(position).count.toString() + ")"
                holder.itemView.setOnClickListener {
                    FolderProvider.instance.selectedFolder = folders.get(position)
                    contentView!!.tvCenter.text = folders.get(position).name
                    popupWindow.dismiss()
                    listener?.OnFolderSelected()
                }
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {
            return FolderHolder(context.getView(R.layout.item_folder, parent))
        }

        override fun getItemCount(): Int {
            return folders.size
        }
    }

    fun isShowing(): Boolean {
        return popupWindow.isShowing
    }

    class FolderHolder(view: View) : RecyclerView.ViewHolder(view)

    interface OnFolderSelectedListener {
        fun OnFolderSelected()
    }
}
